main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,255 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        Expression.ApplyTail(sym, formals, optimizeExps(actuals), tpe, loc)
+      case Expression.ApplyHook(hook, args, tpe, loc) => Expression.ApplyHook(hook, optimizeExps(args), tpe, loc)
+      case Expression.Apply(exp, args, tpe, loc) => Expression.Apply(optimize(exp), optimizeExps(args), tpe, loc)
+      case Expression.Unary(op, exp, tpe, loc) => Expression.Unary(op, optimize(exp), tpe, loc)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) => optimizeBinaryExpression(e0.asInstanceOf[Expression.Binary])
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+        // Elimination of dead branches (e.g. if (true) e1 else e2).
+        val e1 = optimize(exp1)
+        e1 match {
+          case Expression.True => optimize(exp2)
+          case Expression.False => optimize(exp3)
+          case _ =>
+            val e2 = optimize(exp2)
+            val e3 = optimize(exp3)
+            Expression.IfThenElse(e1, e2, e3, tpe, loc)
+        }
+      case Expression.Let(sym, exp1, exp2, tpe, loc) => Expression.Let(sym, optimize(exp1), optimize(exp2), tpe, loc)
+      case Expression.LetRec(sym, exp1, exp2, tpe, loc) => Expression.LetRec(sym, optimize(exp1), optimize(exp2), tpe, loc)
+      // TODO: Elimination of run-time tag checks of singleton-valued enums.
+      case Expression.Is(sym, tag, exp, loc) => Expression.Is(sym, tag, optimize(exp), loc)
+      // TODO: Remove the tag and untag on a single case enum.
+      case Expression.Tag(sym, tag, exp, tpe, loc) => Expression.Tag(sym, tag, optimize(exp), tpe, loc)
+      case Expression.Untag(sym, tag, exp, tpe, loc) => Expression.Untag(sym, tag, optimize(exp), tpe, loc)
+      case Expression.Index(base, offset, tpe, loc) => Expression.Index(optimize(base), offset, tpe, loc)
+      case Expression.Tuple(elms, tpe, loc) => Expression.Tuple(optimizeExps(elms), tpe, loc)
+      case Expression.Existential(fparam, exp, loc) => Expression.Existential(fparam, optimize(exp), loc)
+      case Expression.Universal(fparam, exp, loc) => Expression.Universal(fparam, optimize(exp), loc)
+      case Expression.NativeConstructor(constructor, args, tpe, loc) =>
+        Expression.NativeConstructor(constructor, optimizeExps(args), tpe, loc)
+      case Expression.NativeField(field, tpe, loc) => Expression.NativeField(field, tpe, loc)
+      case Expression.NativeMethod(method, args, tpe, loc) =>
+        Expression.NativeMethod(method, optimizeExps(args), tpe, loc)
+      case Expression.UserError(tpe, loc) => Expression.UserError(tpe, loc)
+      case Expression.MatchError(tpe, loc) => Expression.MatchError(tpe, loc)
+      case Expression.SwitchError(tpe, loc) => Expression.SwitchError(tpe, loc)
+    }
+
+    // Start the timer.
+    val t = System.nanoTime()
+
+    // Optimize expressions in the definitions.
+    val optDefinitions = root.definitions.mapValues(defn => defn.copy(exp = optimize(defn.exp)))
```

Comment:
Map values is horribly broken. It is lazy which is something you might not expect and that can wreck havoc later on. It might be the cause of your issues. Use some form of a fold instead.

I will get back with more feedback :)